### PR TITLE
fix(product, types, workflows): Update product variant workflow

### PR DESCRIPTION
**Pull request by pKorsholm**
**What**
- Fix issues with update-variant workflow: 
  - other variants than the updated variant are no longer removed 
  - options are updated properly

### CODE REVIEW

File: packages/types/src/product/service.ts [hunk 1]
```diff
@@ -1489,87 +1490,92 @@ export interface IProductModuleService {
     sharedContext?: Context
   ): Promise<ProductVariantDTO[]>
 
+  updateVariants(
```
Comment (pKorsholm) [inline packages/types/src/product/service.ts:1494]: Change is here

File: packages/types/src/product/common.ts [hunk 2]
```diff
@@ -226,17 +226,17 @@ export interface ProductVariantDTO {
   width?: number | null
   /**
    * The associated product options.
-   * 
+   *
    * @expandable
    */
-  options: ProductOptionValueDTO
+  options: ProductOptionValueDTO[]
```
Comment (pKorsholm) [inline packages/types/src/product/common.ts:232]: here's the change

File: integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts [hunk 3]
```diff
@@ -250,4 +253,121 @@ describe("[Product & Pricing Module] POST /admin/products/:id/variants/:id", ()
...
+      adminHeaders
+    )
+
+    const response = await api.get(
+      `/admin/products/${product.id}`,
+      adminHeaders
+    )
+
+    expect(response.status).toEqual(200)
+    expect(response.data.product).toEqual(
```
Comment (riqwan) [inline integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts:281]:
```suggestion
    expect(response.status).toEqual(200)
   
 expect(response.data.product.variants[0].options).toHaveLength(1)
   
 expect(response.data.product.variants[1].options).toHaveLength(1)
    expect(response.data.product).toEqual(
```

File: integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts [hunk 4]
```diff
@@ -250,4 +253,121 @@ describe("[Product & Pricing Module] POST /admin/products/:id/variants/:id", ()
...
+      data,
+      adminHeaders
+    )
+
+    const response = await api.get(
+      `/admin/products/${product.id}`,
+      adminHeaders
+    )
+
+    expect(response.status).toEqual(200)
```
Comment (riqwan) [inline integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts:280]:
```suggestion
    expect(response.status).toEqual(200)
    expect(response.data.product.variants[0].options).toHaveLength(1)
    expect(response.data.product.variants[1].options).toHaveLength(1)
```

File: packages/product/src/repositories/product-option-value.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    context: Context = {}
+  ): Promise<ProductOptionValue[]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+    const findOptions_ = { ...findOptions }
+
+    findOptions_.options ??= {}
+
+    Object.assign(findOptions_.options, {
+      strategy: LoadStrategy.SELECT_IN,
+    })
```
Comment (riqwan) [inline packages/product/src/repositories/product-option-value.ts]: This is no longer needed, its a legacy check

```suggestion
```

File: packages/product/src/repositories/product-option-value.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  async upsert(
+    optionValues: (UpdateProductOptionValueDTO | CreateProductOptionValueDTO)[],
+    context: Context = {}
+  ): Promise<ProductOptionValue[]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+
+    const optionValueIds = optionValues
+      .filter((option) => !!option.id)
```
Comment (riqwan) [inline packages/product/src/repositories/product-option-value.ts]:
```suggestion
      .filter((option) => option.id)
```

File: packages/product/src/services/product-module-service.ts [hunk 7]
```diff
@@ -231,6 +244,108 @@ export default class ProductModuleService<
...
+    const optionsToUpsert: (
+      | CreateProductOptionValueDTO
+      | UpdateProductOptionValueDTO
+    )[] = []
+    const optionsToDelete: string[] = []
+
+    const toUpdate = data.map(({ id, options, ...rest }) => {
+      const variant = variantsMap.get(id)!
+
+      const toUpdate: UpdateProductVariantDTO & { product_id: string } = {
```
Comment (riqwan) [inline packages/product/src/services/product-module-service.ts]: should we put this in the types in the product module?

File: packages/product/src/services/product-module-service.ts [hunk 8]
```diff
@@ -231,6 +244,108 @@ export default class ProductModuleService<
...
+      if (options?.length) {
+        const optionIdToUpdateValueMap = new Map(
+          options.map(({ option, option_id, value }) => {
+            const computedOptionId = option_id ?? option.id ?? option
+            return [computedOptionId, value]
+          })
+        )
+
+        for (const existingOption of variant.options) {
+          if (!optionIdToUpdateValueMap.has(existingOption.option)) {
```
Comment (riqwan) [inline packages/product/src/services/product-module-service.ts]:
```suggestion
          if (!optionIdToUpdateValueMap.has(existingOption.id)) {
```

File: packages/product/src/services/product-module-service.ts [hunk 9]
```diff
@@ -231,6 +244,108 @@ export default class ProductModuleService<
...
+            optionsToDelete.push(existingOption.id)
+          } else {
+            optionsToUpsert.push({
+              id: existingOption.id,
+              option: existingOption.option as unknown as string,
```
Comment (riqwan) [inline packages/product/src/services/product-module-service.ts]:
```suggestion
```

ignore if its needed, but seems redundant.

File: packages/product/src/services/product-module-service.ts [hunk 10]
```diff
@@ -231,6 +244,108 @@ export default class ProductModuleService<
...
+              value: optionIdToUpdateValueMap.get(existingOption.option)!,
+            })
+            optionIdToUpdateValueMap.delete(existingOption.option)
```
Comment (riqwan) [inline packages/product/src/services/product-module-service.ts]:
```suggestion
            optionIdToUpdateValueMap.delete(existingOption.id)
```

File: packages/product/src/services/product-module-service.ts [hunk 11]
```diff
@@ -231,6 +244,108 @@ export default class ProductModuleService<
...
+        }
+      }
+
+      if (Object.keys(rest).length) {
+        for (const [key, value] of Object.entries(rest)) {
+          if (variant[key] !== value) {
+            toUpdate[key] = value
+          }
+        }
+      }
```
Comment (riqwan) [inline packages/product/src/services/product-module-service.ts]:
```suggestion
      for (const [key, value] of Object.entries(rest)) {
        if (variant[key] !== value) {
          toUpdate[key] = value
        }
      }
```

the if seems redundant

File: packages/product/src/services/product-module-service.ts [hunk 12]
```diff
@@ -231,6 +244,108 @@ export default class ProductModuleService<
...
+
+      return toUpdate
+    })
+
+    const groups = groupBy(toUpdate, "product_id")
+
+    const [, , result]: [void, TProductOptionValue[], TProductVariant[][]] =
```
Comment (riqwan) [inline packages/product/src/services/product-module-service.ts]:
```suggestion
    const [_, _optionValues, productVariants]: [void, TProductOptionValue[], TProductVariant[][]] =
```

File: packages/product/src/services/product-module-service.ts [hunk 13]
```diff
@@ -231,6 +244,108 @@ export default class ProductModuleService<
...
+            return await this.productVariantService_.update(
+              product_id,
+              update.map(({ product_id, ...update }) => update),
+              sharedContext
+            )
+          })
+        ),
+      ])
+
+    return JSON.parse(JSON.stringify(result.flat()))
```
Comment (riqwan) [inline packages/product/src/services/product-module-service.ts]:
```suggestion
    return await this.baseRepository_.serialize<
      ProductTypes.ProductVariantDTO[]
    >(productVariants, {
      populate: true,
    })
```

**Review by riqwan — APPROVED** [review]
CLEAN!

File: integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts (hunk 4 above)
Comment (pKorsholm) [reply integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts:280]: this is redundant since we're not using `expect.arrayContaining`

File: integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts [hunk 14]
```diff
@@ -250,4 +253,195 @@ describe("[Product & Pricing Module] POST /admin/products/:id/variants/:id", ()
       })
     )
   })
+
+  it("Should update variant option value", async () => {
```
Comment (adrien2p) [inline integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts]: comment: you can lower case all the S of the should:D

File: packages/product/src/repositories/product-option-value.ts [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+  async find(
+    findOptions: DAL.FindOptions<ProductOptionValue> = { where: {} },
+    context: Context = {}
+  ): Promise<ProductOptionValue[]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+    const findOptions_ = { ...findOptions }
+
+    findOptions_.options ??= {}
+
```
Comment (adrien2p) [inline packages/product/src/repositories/product-option-value.ts:31]: comment: rm extra line

File: packages/product/src/repositories/product-option-value.ts [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+        optionValuesToUpdate.push(updatedOptionValue)
+      } else {
+        const newOptionValue = manager.create(ProductOptionValue, {
+          option: option_id,
+          variant: (optionValue as CreateProductOptionValueDTO).variant_id,
+          ...optionValue,
+        })
+        optionValuesToCreate.push(newOptionValue)
+      }
+    })
```
Comment (adrien2p) [inline packages/product/src/repositories/product-option-value.ts]:
```suggestion
optionValues.forEach(({ option_id, ...optionValue }) => {
  const existingOptionValue = existingOptionValuesMap.get(optionValue.id)
  if (optionValue.id && existingOptionValue) {
    const updatedOptionValue = manager.assign(existingOptionValue, {
      option: option_id,
      ...optionValue,
    })
    optionValuesToUpdate.push(updatedOptionValue)
    return
  }
  
  const newOptionValue = manager.create(ProductOptionValue, {
    option: option_id,
    variant: (optionValue as CreateProductOptionValueDTO).variant_id,
    ...optionValue,
  })
  optionValuesToCreate.push(newOptionValue)
})
```

File: packages/product/src/services/product-module-service.ts [hunk 17]
```diff
@@ -231,6 +244,119 @@ export default class ProductModuleService<
...
+    if (variants.length !== data.length) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "Cannot update non-existing variants with ids: " +
+          data
+            .filter(({ id }) => !variantsMap.has(id))
+            .map(({ id }) => id)
+            .join(", ")
+      )
+    }
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts]: question: I think riqwan created a diff utils for that? can you check with him please

File: packages/product/src/services/product-module-service.ts [hunk 18]
```diff
@@ -231,6 +244,119 @@ export default class ProductModuleService<
...
+          options.map(({ option, option_id, value }) => {
+            const computedOptionId = option_id ?? option.id ?? option
+            return [computedOptionId, value]
+          })
+        )
+
+        for (const existingOptionValue of variant.options) {
+          if (!optionIdToUpdateValueMap.has(existingOptionValue.option.id)) {
+            optionsValuesToDelete.push(existingOptionValue.id)
+          } else {
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts]: suggestion: continue, then you can remove the else and just have the block after the if

File: packages/product/src/services/product-module-service.ts [hunk 19]
```diff
@@ -231,6 +244,119 @@ export default class ProductModuleService<
...
+          return await this.productVariantService_.update(
+            product_id,
+            update.map(({ product_id, ...update }) => update),
+            sharedContext
+          )
+        })
+      ),
+    ])
+
+    return await this.baseRepository_.serialize<
```
Comment (adrien2p) [inline packages/product/src/services/product-module-service.ts]: todo: here you are inside the transaction still, maybe create the public method updateVariants and put the update in updateVariants_. that way we are serializing the data and have all the ids etc once the transaction is committed, also don't make the transaction longer than needed

File: packages/workflows/src/handlers/product/update-product-variants.ts [hunk 20]
```diff
@@ -14,21 +14,22 @@ export async function updateProductVariants({
...
+  const updateVariantsData: ProductTypes.UpdateProductVariantOnlyDTO[] = []
   const productModuleService: ProductTypes.IProductModuleService =
     container.resolve(ModulesDefinition[Modules.PRODUCT].registrationName)
 
-  for (const [productId, variantsData = []] of productVariantsMap) {
-    updateProductsData.push({
-      id: productId,
-      variants: variantsData,
-    })
+  for (const [product_id, variantsData = []] of productVariantsMap) {
```
Comment (adrien2p) [inline packages/workflows/src/handlers/product/update-product-variants.ts]: question: the variants data doesnt include the product id already?

Comment (pKorsholm) [reply packages/workflows/src/handlers/product/update-product-variants.ts]: variantsData is the update for the variant, not the variant from the db, I will rename it for clarity 😄

File: .changeset/young-items-drop.md [hunk 21]
```diff
@@ -1,9 +1,10 @@
 ---
 "@medusajs/workflows": patch
 "@medusajs/product": patch
+"@medusajs/pricing": patch
```
Comment (adrien2p) [inline .changeset/young-items-drop.md:4]: todo: There is two changeset, the first one seams targetting the right package, but not this one

File: .github/workflows/action.yml [hunk 22]
```diff
@@ -125,7 +125,6 @@ jobs:
         env:
           DB_PASSWORD: postgres
           DB_USERNAME: postgres
```
Comment (adrien2p) [inline .github/workflows/action.yml:127]: question: Why is this deleted?

File: integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts [hunk 23]
```diff
@@ -250,4 +253,231 @@ describe("POST /admin/products/:id/variants/:id", () => {
...
+                value: "test 2",
+              }),
+            ],
+          }),
+        ]),
+      })
+    )
+  })
+
+  it.only("should update variant metadata", async () => {
```
Comment (adrien2p) [inline integration-tests/plugins/__tests__/product/admin/update-product-variant.spec.ts]: todo: rm only

File: packages/product/src/repositories/product-option-value.ts [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+
+  async upsert(
+    optionValues: (UpdateProductOptionValueDTO | CreateProductOptionValueDTO)[],
+    context: Context = {}
+  ): Promise<ProductOptionValue[]> {
+    const manager = this.getActiveManager<SqlEntityManager>(context)
+
+    const optionValueIds = optionValues
+      .filter((option) => !!option.id)
+      .map(({ id }) => id!)
```
Comment (adrien2p) [inline packages/product/src/repositories/product-option-value.ts]: todo: just thinking that maybe instead of a double loop we could do something like
```ts
const optionValueIds = []
optionValues.forEach(optionValue => {
  if (!!optionValue.id) optionValueIds.push(optionValue.id)
})
```

**Review by adrien2p — APPROVED** [review]

File: .changeset/young-items-drop.md (hunk 21 above)
Comment (pKorsholm) [reply .changeset/young-items-drop.md:4]: `pricing` was missing in the core integration changeset from the pr we merged earlier, just adding it here

File: .github/workflows/action.yml (hunk 22 above)
Comment (pKorsholm) [reply .github/workflows/action.yml:127]: as discussed `split-tests` doesn't exist as a step anywhere
